### DISCUSSION TITLE
[Feedback] Database Refactor Alpha

### DISCUSSION THREAD

**Discussion by kodumbeats** [appwrite/appwrite#2451] [General]
The Appwrite Database refactor is finally ready for an Alpha release and testing! We've compiled docs and SDKs because we need _your_ help to make the Database Service the best it can be. 

Check out our Getting Started guide to learn more:

https://warm-tray-285.notion.site/Appwrite-0-12-Alpha-The-Database-Refactor-70f2c1d234394d9ab9d0d548fe5449bc

> Note: this build is not production-ready and should not be used with sensitive/important data.

To install:
```
git clone --branch db-alpha https://github.com/appwrite/appwrite.git 
cd appwrite 
docker-compose build # Takes a few minutes to compile Appwrite from source 
docker-compose pull # Pulls all other required images 
docker-compose up -d # Start the Appwrite stack
```

Leave all your questions, comments, concerns, and bug reports here in this discussion. We'll collect the most frequent questions in Discord and in this thread and create a Frequently Asked Questions (FAQs) as we approach the full release.

Thanks for being the best community 🙏
~ @kodumbeats

**Comment by outrelouxhe** [comment]
Hello, when I insert more than 5000 documents in a collection,  the console shows  `5000+ documents found` and the number of pages of 12 documents is limited to 417 (417 * 12 = 5004).  On the API side, the listDocuments returns a `sum` value of exactly 5000 event if there are 25000+ documents.  Is it a limit of the actual alpha version?

**Reply by kodumbeats** [reply]
This is an intentional design decision to protect the database, and we have a great real-world example with the same: GitHub Stargazers. First, some background:

Counting all records in a table/collection is an expensive operation on the database because it must scan through every single row (e.g. read the data from disk) to get an accurate answer. Instead, databases opt to provide an estimated count that's much easier for the database to collect. Additionally, we rarely send 25k+ documents to a client - instead, apps implement some kind of pagination to get the data in chunks. 

Since Appwrite is designed to host many projects on one database, we must be very careful of the database performance - it's easy to scale stateless Docker containers but difficult to scale stateful databases. 

Back to Stargazers - `appwrite/appwrite` has over 14k Stargazers, while `appwrite/awesome-appwrite` has a smaller, more accurate count. In both cases, the Stargazer page displays only 30 results:

https://github.com/appwrite/appwrite/stargazers
https://github.com/appwrite/awesome-appwrite/stargazers

After 5000 Stargazers, GitHub stops providing an accurate count and employs [cursor pagination](https://stackoverflow.com/questions/18314687/how-to-implement-cursors-for-pagination-in-an-api) to maintain performance as you navigate through the pages.

A `sum` is necessary `limit+offset` pagination, which, until the refactor, was the only way to paginate from the Database API. Now, the API also supports `cursor` pagination, so developers can take advantage of the massive performance improvements between the two techniques.

**Reply by outrelouxhe** [reply]
Thank you for your reply that makes a lot of sense as I also experience too long response time on  queries on large collections on the 0.11 version.
So what are the solutions if we want to know anyway the number of documents in a large collection? I can think about  a cloud function triggered by document creation/deletion that will keep track of the number of documents and store the actual count in a document of a dedicated collection. Is it a good practice? Do you have other ideas in mind? Thank you

**Reply by kodumbeats** [reply]
Your cloud function solution sounds pretty solid, though it may not scale very well - you might eventually take advantage of `cursor` pagination and execute the function on a schedule instead of every request. Data aggregation in this manner is pretty common for very large tables (millions of rows)

**Comment by jbdujardin** [comment]
Hello.
Why there is not a DateTime attribute?
Do You suggest to use String with DateTime ISO format or an integer with the timestamp?
Thank you for this very good work.

**Reply by kodumbeats** [reply]
Great question!

We've considered a dedicated DateTime validator, but we couldn't decide if it should be stored as some ISO8601 string or UNIX timestamp - there's [advantages to both](https://stackoverflow.com/a/178818).

What is your preference?

**Reply by lohanidamodar** [reply]
I actually prefer UNIX timestamp, it's lot more easier to query than counterpart string format. Would love to have a timestamp attribute.

Also is it possible to have both attribute - Timestamp as UNIX timestamp and DateTime as ISO8601 string?

**Reply by jbdujardin** [reply]
I prefer UNIX Timestamp too.
It’s more easier to sort. 
We could use integer so no attribute needed I think.

**Reply by stnguyen90** [reply]
I think RFC 3339 is best because https://stackoverflow.com/questions/10286204/what-is-the-right-json-date-format and I believe golang also uses that format as the default for encoding/decoding JSON.

**Reply by lohanidamodar** [reply]
> I prefer UNIX Timestamp too. It’s more easier to sort. We could use integer so no attribute needed I think.

We could use int but there would be not validation. We want it for validation (for example for URL, Email, we could use use string, but we have those for easier validation and data integrity)

**Comment by ewhanson** [comment]
Hi, are there any plans to include something like a JSON attribute type to replace the deprecated wildcard rule?

**Reply by eldadfux** [reply]
We're still looking into that and researching other ways to support relationships. Adding a Graph DB adapter has been an idea we have played with. For now the suggested way is to use a plain string for storing JSON data.

**Comment by stnguyen90** [comment]
I tried to set a read access to a team (`team:61b068162c0372d13b8a`) via the console, but got the following error:

```json
{
    "message": "Read permissions must be one of: (1, 1, 1, 1)",
    "code": 400,
    "version": "db-alpha",
    "file": "/usr/src/code/app/controllers/api/database.php",
    "line": 2005,
    "trace": [
        {
            "file": "/usr/src/code/vendor/utopia-php/framework/src/App.php",
            "line": 582,
            "function": "{closure}",
            "args": [
                "61b064ca6fb055cb46f3",
                "61b06847b6485b2df46d",
                {
                    "name": "Jessica Alba",
                    "$id": "61b06847b6485b2df46d",
                    "$internalId": "1",
                    "$read": [
                        "team:61b068162c0372d13b8a"
                    ],
                    "$write": [],
                    "$collection": "61b064ca6fb055cb46f3"
                },
                "team:61b068162c0372d13b8a",
                [],
                {},
                {},
                {},
                {},
                {}
            ]
        },
        {
            "file": "/usr/src/code/vendor/utopia-php/framework/src/App.php",
            "line": 683,
            "function": "execute",
            "class": "Utopia\\App",
            "type": "->",
            "args": [
                {},
                {
                    "data": {
                        "name": "Jessica Alba"
                    },
                    "read": [
                        "team:61b068162c0372d13b8a"
                    ],
                    "write": []
                }
            ]
        },
        {
            "file": "/usr/src/code/app/http.php",
            "line": 185,
            "function": "run",
            "class": "Utopia\\App",
            "type": "->",
            "args": [
                {},
                {}
            ]
        }
    ]
}
```

Any ideas?

**Reply by stnguyen90** [reply]
Btw, this happens when updating the permission of a Document for a Collection that has Collection level permissions. Adding `team:team:61b068162c0372d13b8a` to the Collection permission works fine.

**Reply by TorstenDittmann** [reply]
Interesting, let me try and reproduce it 🙏

**Reply by stnguyen90** [reply]
Typo with my previous comment. Adding `team:61b068162c0372d13b8a` to the collection permission works, but not document permission.

**Reply by TorstenDittmann** [reply]
Thank you, was able to reproduce it really easy 👍🏻

**Reply by TorstenDittmann** [reply]
Found it and fixed 🙏🏻  https://github.com/appwrite/appwrite/pull/2477

It was an error with checking the permissions using an Admin User (the user of the console). Also the error message was fixed to no show a few `1`'s 👍🏻

**Comment by stnguyen90** [comment]
It seems my app broke because of API changes (`$read` and `$write` are no longer nested). When this update is released, will it be backwards compatible with older SDKs/apps?

**Reply by christyjacob4** [reply]
@stnguyen90 If you're talking about the response we do offer backwards compatibility. We have a concept of **Response Filters** that allow your SDKs to request for an older Appwrite version's response. Once we go out of alpha and release 0.12, backwards compatibility should work well 👍 

Passing the `X-Appwrite-Response-Format: 0.11.0` will return responses according to Appwrite 0.11.0  
Like this https://github.com/appwrite/sdk-for-flutter/blob/master/lib/src/client_browser.dart#L37

So as long as you don't upgrade your SDK version, you should be good.

**Comment by TheoLeeCJ** [comment]
I might have encountered a possible bug with the Flutter SDK for AppWrite. I apologise in advance if the bug cannot be reproduced.

I followed the instructions here to install the alpha SDK: https://warm-tray-285.notion.site/Flutter-Dart-805181285c0e45259c12a521a643a0b0.

```yaml
  dart_appwrite:
    git:
      url: https://github.com/appwrite/sdk-for-dart
      ref: db-alpha
  appwrite:
    git:
      url: https://github.com/appwrite/sdk-for-flutter
      ref: db-alpha
```

And then I deployed an AppWrite 0.12 instance using the instructions on the Notion document. Afterwards, I attempted to use the Account service.

```dart
import 'package:appwrite/appwrite.dart';

void main() {
  Client client = Client();
  client.setEndpoint("https://...").setProject("...");
  late Account appwriteAccountService = Account(client);

  Timer(const Duration(milliseconds: 5000), () { // just for testing
    appwriteAccountService.get()
      .then((user) { print(user); })
      .catchError((onError) { print("at init: "); print(onError); });
  });

  runApp(const App());
}
```

This subsequently gave the following error:

```
E/flutter (18777): [ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: type 'List<dynamic>' is not a subtype of type 'Map<String, dynamic>'
E/flutter (18777): #0      new User.fromMap (package:appwrite/src/models/user.dart:42:43)
E/flutter (18777): #1      Account.get (package:appwrite/services/account.dart:21:28)
E/flutter (18777): <asynchronous suspension>
E/flutter (18777): #2      main.<anonymous closure> (package:test/main.dart:26:5)
E/flutter (18777): <asynchronous suspension>
```

This behaviour was not present in the current 2.0.3 version of the Flutter SDK with an AppWrite 0.11 instance.

And here is some information:
v:db-alpha.181 (from AppWrite console)

As of time of writing.

Thank you very much! AppWrite is an amazing project, and I look forward to the release of this database refactor - it seems to improve performance.

**Reply by stnguyen90** [reply]
I think I sort of experienced this too. I didn't test with flutter, but I did notice the `account.get()` API call returned the user preferences as `[]` rather than the expected `{}` which would cause the `type 'List<dynamic>' is not a subtype of type 'Map<String, dynamic>'` error in dart/flutter. I'm not sure if it makes any difference, but I created my user via the admin console.

**Reply by TorstenDittmann** [reply]
Found the issue, the default value was an empty array when a user was created 👍🏻 

This got resolved now in the dev branch 😃

**Comment by BBurnworth** [comment]
I installed the Alpha branch on Digital Ocean VPS - Ubuntu 20 in the same folder where i had .11 previously installed and I was getting an Error 500 when accessing from an outside network (previous .11 was working correctly). Thanks to Eldad and Drake on discord for troubleshooting with me.

I wonder if the issue was a database password conflict between versions.

**Note**: This clears volumes which may result in losing data. 

**To Fix:** 

I had to set the .env variables for APP_DOMAIN, and APP_TARGET_DOMAIN, and also following this: [Debugging Appwrite SSL Certificates](https://gist.github.com/eldadfux/eb3ed1c4e5f43b7f7259469dd29312c5) I set the ENV & SECURITY_EMAIL settings. 

Also, I had to run `docker-compose down -v` in the appwrite folder and then run `docker-compose build`, `docker-compose pull`, `docker-compose up -d` before it would connect from the outside.

**Reply by eldadfux** [reply]
Looking at some of the logs you shared on Discord, this seems to me like a conflict in the database password between the versions you had installed. Happy the solution worked for you.

**Comment by kodumbeats** [comment]
Hey y'all!

One of our pending questions that we could use your help to answer: **what's the future of pagination?**

In the current state of the Database API, the `listDocuments` endpoint supports both `limit+offset` and `cursor` pagination. Neither suit 100% of use cases:

- `limit+offset` pagination is easier to understand and implement, but does not scale well as the collection/offset grows.
- `cursor` pagination scales very well, but is not the easiest to explain or implement.

## What do you think - should we keep both pagination schemes, or focus on just one? 

Prior art:
- https://stackoverflow.com/questions/18314687/how-to-implement-cursors-for-pagination-in-an-api
- https://stackoverflow.com/questions/55744926/offset-pagination-vs-cursor-pagination

**Reply by kodumbeats** [reply]
My opinion:
We bias towards performance in other areas of the Database API (maximum document limit as an example), so I believe the same should apply here. I think we should only support cursor pagination, but we should invest in detailed documentation and examples to make it easier to understand and integrate into apps.

**Reply by eldadfux** [reply]
I tend to be with your on this. Our Database is designed for performance and our API is designed for simplicity. Both work better with cursor pagination as a sole option. I think explaining how to work with two options at parallel under different limitation would be harder than explaining the cursor pagination.

**Reply by stnguyen90** [reply]
It would be nice to also support limit + offset since there are some libraries that primarily support that:

- [material-table](https://material-table.com/#/docs/features/remote-data)
- [react-table](https://react-table.tanstack.com/docs/api/usePagination)

As for which to focus on, would any choice impact/delay the release?

**Reply by Meldiron** [reply]
Storytime, when working with GitHub API that supports pagination (page=number), I noticed their limit is per_page=100 and page=10. If you try to get the 11th page, you get an error.

A workaround was to split my query to multiple queries, so I don't have more than 10 pages (instead of loading the whole month in one query, I load each day separately). Another solution would be to use cursor pagination.

@kodumbeats Do you think we can implement it similarly?

**Comment by Meldiron** [comment]
Our current implementation does not allow ORs on different attributes. In MySQL world it would be `WHERE name = 'Matej' OR age = 12`. 

It is just an edge-case, I would like to see if anyone would need this.

**Reply by gepd** [reply]
I'm currently working in a proyect where I need this king of implementation. 

This is specially usefull when when you need filter data in an ecommerce project (filter by size, color, etc)

**Reply by stnguyen90** [reply]
> This is specially usefull when when you need filter data in an ecommerce project (filter by size, color, etc)

I would epect this type of filtering to be size is S or M AND color is black or blue rather than size is S or M OR color is black or blue. Which is it you would need?

**Reply by tannermeade** [reply]
What is the reasoning behind forcing the AND operation between all the query filters?

If Appwrite doesn't plan to support it, I can see people having a hard time trusting the db. An OR operation between different attributes seems like a pretty basic need.

Like performance is great -- but when there isn't a BigO algorithm of 1 available, that doesn't mean basic forms of expression should be thrown out. There should be options available here for achieving a log(n) class algorithm to get basic forms of boolean expressions.

**Reply by stnguyen90** [reply]
> What is the reasoning behind forcing the AND operation between all the query filters?

I'd say it's pretty common in REST APIs for filters to be combined like AND. Even [firebase](https://cloud.google.com/firestore/docs/query-data/queries#query_limitations) and [supabase](https://github.com/supabase/postgrest-js/issues/44) behave like this.

**Reply by tannermeade** [reply]
I could be wrong, but I'm pretty sure with an OR between attributes Appwrite gains functional completeness where all boolean expressions can be represented.

Providing a NAND operator between attributes could also give functional completeness. This is logically different from an AND paired with the opposite expression that is currently possible.

**Reply by tannermeade** [reply]
> > What is the reasoning behind forcing the AND operation between all the query filters?
> 
> I'd say it's pretty common in REST APIs for filters to be combined like AND. Even [firebase](https://cloud.google.com/firestore/docs/query-data/queries#query_limitations) and [supabase](https://github.com/supabase/postgrest-js/issues/44) behave like this.

To be honest, I really don't want Appwrite to be like Supabase. They've built their whole platform on the success of postgreSQL and postgREST which still can't work well on arm64.

**Comment by finiti** [comment]
Seem to have identified a bug with realtime and database collection permissions.

Collection level read is set as seen below.
"$id": "collectionname",
    "$read": [
        "role:all"
    ],
    "$write": [],
    "name": "collectionname",
    "permission": "collection",


Inserted documents with permissions below do not get triggered on update.
"$id": "documentid",
    "$internalId": "3740",
    "$read": [],
    "$write": [],
    "$collection": "collectionname" 

Inserted documents with permissions below get triggered on update.
"$id": "documentid",
    "$internalId": "3740",
    "$read": [
        "role:all"
    ],
    "$write": [],
    "$collection": "collectionname" 


It appears for realtime the collection level read does not get applied.

**Reply by stnguyen90** [reply]
I can confirm this behavior as well. ListDocuments returns the document, but it won't show up in realtime unless explicitly given read permissions.

**Reply by TorstenDittmann** [reply]
Good catch! Fixing it 👀

**Comment by stnguyen90** [comment]
It seems like setting function variables seems to not work in the db-alpha. Debugging the Appwrite console shows the hidden input isn't being updated when the inputs are updated:

![image](https://user-images.githubusercontent.com/1477010/146617651-df8e7de3-ee96-4314-a593-b45746321abc.png)

**Reply by TorstenDittmann** [reply]
Pretty sure this is fixed on the Dev branch already 👀

Gonna verify asap

**Reply by BBurnworth** [reply]
I am having the same issue.

**Comment by BBurnworth** [comment]
Creating a float attribute with default value 0 fails and the log shows this:

`Default value 0 does not match given type double`

Also, using 0.0 does not work. There is no explanation on the attribute creator page to identify this as the problem.

**Reply by stnguyen90** [reply]
I just tested and got this same issue. The API request definitely sends a numeric value for the default and not a string, but something seems to be going wrong server side.

**Reply by TorstenDittmann** [reply]
This should be fixed on the dev branch now, the value passed wasnt converted to a float value 👍🏻

**Comment by stnguyen90** [comment]
I just ran into a problem when creating a document with an attribute value (string) bigger than the max size. It seems like the response status code is 200, content type is text/html, and response body is not JSON so the `createDocument()` fails to parse the response:

![image](https://user-images.githubusercontent.com/1477010/147318854-925d3f4c-01ae-4c28-a3cd-3286250e250e.png)

**Reply by eldadfux** [reply]
@stnguyen90 Thanks the team is investigating this error!

**Reply by TorstenDittmann** [reply]
This got also fixed on the dev branch now!

Thank you alot @stnguyen90 - your feedback was is incredibly helpful 👏🏻

**Comment by stnguyen90** [comment]
Is the `db.getAttribute()` API working? I keep getting 404 with it.

```shell
root@b7339f9f8639:/usr/local/src# curl -H "X-Appwrite-Key: key" \
  -H "X-Appwrite-Project: proj" \
  http://192.168.1.41/v1/database/collections/c/attributes | json_pp
{
   "attributes" : [
      {
         "key" : "otherId",
         "size" : 50,
         "array" : false,
         "default" : null,
         "status" : "available",
         "required" : true,
         "type" : "string"
      }
   ],
   "sum" : 1
}
root@b7339f9f8639:/usr/local/src# curl -H "X-Appwrite-Key: key" \
  -H "X-Appwrite-Project: proj" \
  http://192.168.1.41/v1/database/collections/c/attributes/otherId | json_pp
{
   "file" : "/usr/src/code/app/controllers/api/database.php",
   "version" : "db-alpha",
   "line" : 1208,
   "code" : 404,
   "message" : "Attribute not found",
   "trace" : [
      {
         "args" : [
            "c",
            "otherId",
            {},
            {},
            {}
         ],
         "file" : "/usr/src/code/vendor/utopia-php/framework/src/App.php",
         "function" : "{closure}",
         "line" : 582
      },
      {
         "type" : "->",
         "args" : [
            {},
            []
         ],
         "file" : "/usr/src/code/vendor/utopia-php/framework/src/App.php",
         "function" : "execute",
         "line" : 683,
         "class" : "Utopia\\App"
      },
      {
         "function" : "run",
         "line" : 185,
         "class" : "Utopia\\App",
         "file" : "/usr/src/code/app/http.php",
         "type" : "->",
         "args" : [
            {},
            {}
         ]
      }
   ]
}
```

**Reply by eldadfux** [reply]
@stnguyen90 Thanks the team is investigating this error!

**Reply by TorstenDittmann** [reply]
This got fixed now on the dev branch 🥳

**Comment by outrelouxhe** [comment]
Thank you to the team for this amazing new version! Is it planned to release soon a new alpha or beta version with the different fixes made recently? Tx.
